packages/graalvm/src/main/kotlin/elide/runtime/plugins/env/EnvPlugin.kt

```diff
@@ -1,56 +1,52 @@
 package elide.runtime.plugins.env
 
 import org.graalvm.polyglot.Value
-import org.graalvm.polyglot.proxy.*
-import java.util.concurrent.ConcurrentSkipListMap
-import java.util.concurrent.atomic.AtomicReference
+import org.graalvm.polyglot.proxy.ProxyHashMap
+import org.graalvm.polyglot.proxy.ProxyIterable
+import org.graalvm.polyglot.proxy.ProxyIterator
+import org.graalvm.polyglot.proxy.ProxyObject
 import elide.runtime.core.*
-import elide.runtime.core.EngineLifecycleEvent.*
+import elide.runtime.core.EngineLifecycleEvent.ContextCreated
+import elide.runtime.core.EngineLifecycleEvent.ContextInitialized
 import elide.runtime.core.EnginePlugin.InstallationScope
 import elide.runtime.core.EnginePlugin.Key
-import elide.runtime.gvm.internals.GraalVMGuest
-import elide.runtime.gvm.internals.intrinsics.js.struct.map.JsMap
-import elide.runtime.plugins.env.Environment.Plugin.APP_ENV_ACCESSOR
-import elide.runtime.plugins.env.Environment.Plugin.APP_ENV_SYMBOL
 import elide.vm.annotations.Polyglot
 
 /**
  * Engine plugin providing isolated application environment.
  *
+ * In order to correctly support all languages, this plugin is cooperative: a language plugin may opt into
+ * using it by [installing][install] the env map binding into a context.
+ *
  * @see EnvConfig
  */
-@DelicateElideApi public class Environment private constructor (public val config: EnvConfig) {
-  private val effectiveEnvironment: AtomicReference<MutableMap<String, String?>> = AtomicReference(null)
```

Comment:
environment is now mapped from the `config`, we no longer need to manually synchronize or even assign this map (lazy is technically still synchronized).